egs/wsj/s5/steps/nnet3/chain/get_egs.py

```diff
@@ -0,0 +1,727 @@
+        distances = []
+        additional_cost = 10
+        for cw in chunk_widths:
+            distance_from_left_boundary = width % cw
+            distance_from_right_boundary = cw - distance_from_left_boundary
+            if distance_from_left_boundary > distance_from_right_boundary:
+                distance = distance_from_right_boundary
+            else:
+                distance = distance_from_left_boundary + additional_cost
+            distances.append([cw, distance])
+        distances = sorted(distances, key = lambda x:x[1])
+        utt_partition_preference[utt_name] = map(lambda x: x[0], distances)
+
+    utts_per_chunk_width = {}
+    for cw in chunk_widths:
+        utts_per_chunk_width[cw] = []
+
+    # get the initial assignment
+    for utt, preferences in utt_partition_preference.items():
+        utts_per_chunk_width[preferences.pop(0)].append(utt)
+
+    # merge partitions until each of them has atleast frames_per_iter frames
+    for iteration in xrange(len(chunk_widths) - 1):
+        has_min_num_frames = True
+        for chunk_width in chunk_widths:
+            if GetNumFrames(feat_dir, utts_per_chunk_width[chunk_width]) <= frames_per_iter:
+                has_min_num_frames = False
+                # we don't want to use this chunk_width if there are less
+                # than frames_per_iter frames as there will not be sufficient
+                # randomization
+                for utt in utts_per_chunk_width.pop(chunk_width):
+                    preferences = utt_partition_preference[utt]
+                    while len(preferences) > 0:
+                        key = preferences.pop(0)
+                        if utts_per_chunk_width.has_key(key):
+                            utts_per_chunk_width[key].append(utt)
+                            break
+
+        if has_min_num_frames:
+            break
+    return utts_per_chunk_width
+
+def GenerateTrainingExamplesFromUtts(dir, lat_dir, chain_dir, feat_dir,
+                                     utts, chunk_width,
+                                     train_feats_string, train_egs_opts_string,
+                                     supervision_opts, ivector_opts,
+                                     num_jobs, max_shuffle_jobs_run,
+                                     frames_per_iter, cmd, only_shuffle):
+
+    # The examples will go round-robin to egs_list.  Note: we omit the
+    # 'normalization.fst' argument while creating temporary egs: the phase of egs
+    # preparation that involves the normalization FST is quite CPU-intensive and
+    # it's more convenient to do it later, in the 'shuffle' stage.  Otherwise to
+    # make it efficient we need to use a large 'nj', like 40, and in that case
+    # there can be too many small files to deal with, because the total number of
+    # files is the product of 'nj' by 'num_archives_intermediate', which might be
+    # quite large.
+    num_frames = data_lib.GetNumFrames(feat_dir, utts)
```

Comment:
On Mon, Apr 11, 2016 at 1:32 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In egs/wsj/s5/steps/nnet3/chain/get_egs.py
> <https://github.com/kaldi-asr/kaldi/pull/686#discussion_r59275930>:
>
> > +def GenerateTrainingExamplesFromUtts(dir, lat_dir, chain_dir, feat_dir,
> > +                                     utts, chunk_width,
> > +                                     train_feats_string, train_egs_opts_string,
> > +                                     supervision_opts, ivector_opts,
> > +                                     num_jobs, max_shuffle_jobs_run,
> > +                                     frames_per_iter, cmd, only_shuffle):
> > +
> > +    # The examples will go round-robin to egs_list.  Note: we omit the
> > +    # 'normalization.fst' argument while creating temporary egs: the phase of egs
> > +    # preparation that involves the normalization FST is quite CPU-intensive and
> > +    # it's more convenient to do it later, in the 'shuffle' stage.  Otherwise to
> > +    # make it efficient we need to use a large 'nj', like 40, and in that case
> > +    # there can be too many small files to deal with, because the total number of
> > +    # files is the product of 'nj' by 'num_archives_intermediate', which might be
> > +    # quite large.
> > +    num_frames = data_lib.GetNumFrames(feat_dir, utts)
>
> OK, got it.
>
> OK, in that case I will take the --{min,max}-{left,right}-context values
> from the user and randomly sample from these for each archive. Right now I
> am allowing the user to specify each chunk-width manually. So to specify
> three chunk-widths they would basically specify three --chunk-width
> options
>

I think specifying the same option multiple times is very susceptible to
misinterpretation, and against the general Kaldi pattern of key=value
specification of configuration values.  Why not specify the min and max,
like for the context, and have a num-sizes variable with a reasonable
default, like 5, which wouldn't normally have to be modified?  Then if the
user just specifies --frames-per-chunk, have it set both the min and max.


> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/686/files/ee9df73d5347ff02db360f6c317f00a202a3f064#r59275930>
>